### PR TITLE
Add new statistics for knowing when and why connections were closed

### PR DESCRIPTION
**Pull request by pfreixes**
We add 4 new statistics which will be used by the users for knowing when the connections were closed and what was the reason, the new four statistics are:

- `connections_closed_broken`: Total connections that were closed due to be in broken state.
- `connections_closed_invalid`: Total connections that were closed due to be considered invalid.
- `connections_closed_max_lifetime`: Total connections that were closed because they reached the max lifetime.
- `connections_closed_idle_timeout`: Total connections that were closed because they reached the idle timeout.

### CODE REVIEW

File: bb8/src/inner.rs [hunk 1]
```diff
@@ -85,7 +87,7 @@ where
     }
 
     pub(crate) async fn get(&self) -> Result<PooledConnection<'_, M>, RunError<M::Error>> {
-        let mut kind = StatsKind::Direct;
+        let mut kind = StatsGetKind::Direct;
```
Comment (djc) [inline bb8/src/inner.rs:90]: Please put this change in a separate commit or PR.

File: bb8/src/internals.rs [hunk 2]
```diff
@@ -139,22 +143,34 @@ where
         ApprovalIter { num: num as usize }
     }
 
-    pub(crate) fn reap(&mut self, config: &Builder<M>) -> ApprovalIter {
+    pub(crate) fn reap(&mut self, config: &Builder<M>) -> (ApprovalIter, u64, u64) {
```
Comment (djc) [inline bb8/src/internals.rs:146]: Let's split the `reap()` changes out in a separate PR.

File: bb8/src/internals.rs [hunk 3]
```diff
@@ -272,6 +292,28 @@ impl AtomicStatistics {
                 .fetch_add(wait_time.as_micros() as u64, Ordering::SeqCst);
         }
     }
+
+    pub(crate) fn record_connection_closed(&self, kind: StatsConnectionClosedKind) {
```
Comment (djc) [inline bb8/src/internals.rs:296]: This can just be named `record()`.

**Comment by pfreixes** [comment]
Closing this one in favour of https://github.com/djc/bb8/pull/205
